packages/infer-schema/src/transforms/neo4j-graphql/graphql.ts

```diff
@@ -0,0 +1,105 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Neo4jStruct, NodeMap, RelationshipMap } from "../../types";
+import createNodeFields from "./utils/create-node-fields";
+import uniqueString from "../../utils/unique-string";
+import { NodeDirective } from "./directives/Node";
+import { GraphQLNode } from "./GraphQLNode";
+import inferRelationshipPropsName from "./utils/infer-relationship-props-name";
+import { RelationshipPropertiesDirective } from "./directives/RelationshipProperties";
+import createRelationshipFields from "./utils/create-relationship-fields";
+import { ExcludeDirective } from "./directives/Exclude";
+
+type GraphQLNodeMap = {
+    [key: string]: GraphQLNode;
+};
+
+export default function graphqlFormatter(neo4jStruct: Neo4jStruct, readonly = false): string {
+    const { nodes, relationships } = neo4jStruct;
+    const bareNodes = transformNodes(nodes, readonly);
+    const withRelationships = hydrateWithRelationships(bareNodes, relationships);
+    const sorted = Object.keys(withRelationships).sort();
+    return sorted.map((typeName) => withRelationships[typeName].toString()).join("\n\n");
+}
+
+function transformNodes(nodes: NodeMap, readonly: boolean): GraphQLNodeMap {
+    const out = {};
+    const takenTypeNames: string[] = [];
+    Object.keys(nodes).forEach((nodeType) => {
+        const neo4jNode = nodes[nodeType];
+        const mainLabel = neo4jNode.labels[0];
+        const typeName = mainLabel.replace(/[^_0-9A-Z]+/gi, "_");
+        const uniqueTypeName = uniqueString(typeName, takenTypeNames);
+        takenTypeNames.push(uniqueTypeName);
+
+        const node = new GraphQLNode("type", uniqueTypeName);
+        const nodeDirective = new NodeDirective();
+        if (mainLabel !== uniqueTypeName) {
+            nodeDirective.addLabel(mainLabel);
+        }
+        nodeDirective.addAdditionalLabels(neo4jNode.labels.slice(1));
+        if (nodeDirective.toString().length) {
+            node.addDirective(nodeDirective);
+        }
+        if (readonly) {
+            const excludeDirective = new ExcludeDirective();
+            excludeDirective.addOperation("CREATE");
+            excludeDirective.addOperation("DELETE");
+            excludeDirective.addOperation("UPDATE");
+            node.addDirective(excludeDirective);
+        }
+
+        const fields = createNodeFields(neo4jNode.properties, node.typeName);
+        fields.forEach((f) => node.addField(f));
+        if (node.fields.length) {
+            out[mainLabel] = node;
+        }
+    });
+    return out;
+}
+
+function hydrateWithRelationships(nodes: GraphQLNodeMap, rels: RelationshipMap): GraphQLNodeMap {
+    Object.entries(rels).forEach(([realType, rel]) => {
```

Comment:
Just noticed this typo which is causing test failures locally:

```suggestion
    Object.entries(rels).forEach(([relType, rel]) => {
```

Reply:
Ugh, sorry, I just committed @angrykoala's change suggestion from the GH UI.